### ISSUE TITLE
Feature request - Make Templ have a simpler way to access context. 

### ISSUE THREAD

**Issue by gungun974** [a-h/templ#253]
Before 2fcbab7499977b968c2cdf1fd59e8dbec4fbaca2, templ variable name were not random. This could create some issue and I'm happy this is resolve.

But in the issue #169, a decision was made to also make "random" `w` and `ctx` (I'm not sure seeing the code it's really random or just hard coded with a long prefix).

I will not argue to bring back `w` since it's dangerous to mess with the output string. 
No my issue is really just with `ctx` which is something really helpful for situation like passing current URL to a navbar component.

With the vanish to this easy access, it's become hard to just grab the context.

So here some proposal to solve this problem.

1. (obvious) We can bring back `ctx` like before but this could create some collision if someone.
2. We can create a fake method on `templ` runtime like `templ.GetCtx()` which the generator will just replace this with the variable `templ_7745c5c3_Ctx` in the generated code. 
Like this we don't mess with collision but we need to make a fake method on `templ` runtime for LSP completion.

This is for now just a suggestion but if @a-h for example think my second proposal would be interesting. I could try to make a PR this weekend.

**Comment by joerdav** [comment]
From the response it's clear that the fact `ctx` was accessible was being exploited/used more than expected, so maybe a roll back is in order not to break everyone. But like you say it needs some thinking. 


The reason I think accessing the ctx in render should be avoided is because I think it provides some odd looking code. Reading the template it's not obvious where context is being passed from.

I wonder in your use case why it's not possible to add ctx onto the parameters of the `templ` function, this shows that it's explicitly needed by the template.

**Comment by gungun974** [comment]
> From the response it's clear that the fact `ctx` was accessible was being exploited/used more than expected, so maybe a roll back is in order not to break everyone. But like you say it needs some thinking.

Discussions and ideas are key

> The reason I think accessing the ctx in render should be avoided is because I think it provides some odd looking code. Reading the template it's not obvious where context is being passed from.

That's fair. This should not be a pattern that everyone should use everyday. 
It's just go context was made to make easy to pass some occasionally data between a standard interface.

For the fact `ctx` was kind popping for nowhere. I think this problem like I propose could be address by creating a method (real or not) `templ.GetContext` or `templ.GetCtx` who will just return the context.
It's no more a magical variable in the `.templ` files and more clear to read

> I wonder in your use case why it's not possible to add ctx onto the parameters of the `templ` function, this shows that it's explicitly needed by the template.

I have two issue of why I can't add in my `HeaderLink` component a `context.Context`.

1. Here me tree in my project to render the Index Page. `@IndexPage() > @layouts.AppLayout() > @AppHeader() > @HeaderLink()`.
It's long and none of those component except `@HeaderLink()` need context and if I do that I need to make sure my app Context is always pass.
2. This make me do a double context in the generated file and my render function...
```go
func HeaderLink(ctx context.Context, link Link) templ.Component {
	return templ.ComponentFunc(func(templ_7745c5c3_Ctx context.Context, templ_7745c5c3_W io.Writer) (templ_7745c5c3_Err error) {
		...
	})
}

```

**Comment by joerdav** [comment]
I think I understand the use case. You'd like to avoid a `context.Context` parameter on all of the functions in the stack `@IndexPage() > @layouts.AppLayout() > @AppHeader() > @HeaderLink()`.

From my view though it's a common pattern in Go to pass context down through layers if it's needed. So `@IndexPage(ctx context.Context) > @layouts.AppLayout(ctx context.Context) > @AppHeader(ctx context.Context) > @HeaderLink(ctx context.Context)`.

I think the "uglified" templ ctx var was an attempt to encourage this kind of explicit context passing to be closer aligned to a go program.

I appreciate this doesn't solve your current problem though as it would involve some rework on your part.

**Comment by a-h** [comment]
I touched on my thoughts here - https://github.com/a-h/templ/issues/169#issuecomment-1775126500

Just to further explain, with the "implicit context" pattern (there's a ctx variable that just happens to exist in templ components), so templates can use standard HTTP middleware without any drama.

```go
type contextKey string
type contextClass = contextKey("class")

func Middleware(next http.Handler) http.Handler {
  return http.HandlerFunc(func (w http.ResponseWriter, r *http.Request ) {
    ctx := context.WithValue(r.Context(), contextClass, "red")
    next.ServeHTTP(w, r.WithContext(ctx))
  })
}

templ Page() {
  @Show()
}

templ Show() {
  <div class={ ctx.Value(contextClass) }>Display</div>
}

func main() {
  h := templ.Handler(Page())
  withMiddleware := Middleware(h)
  http.Handle("/", withMiddleware)
  http.ListenAndServe(":8080", h)
}
```

With the "explicit context" pattern, you have to wrap templ components in a handler to take the request context and put it into the template constructor function call, plus drill the context through all of your components from the top level component through to all children.

```go
type contextKey string
type contextClass = contextKey("class")

func Middleware(next http.Handler) http.Handler {
  return http.HandlerFunc(func (w http.ResponseWriter, r *http.Request ) {
    ctx := context.WithValue(r.Context(), contextClass, "red")
    next.ServeHTTP(w, r.WithContext(ctx))
  })
}

templ Page(ctx context.Context) {
  @Show(ctx)
}

templ Show(ctx context.Context) {
  <div class={ ctx.Value(contextClass) }>Display</div>
}

func main() {
  h := http.HandlerFunc(func (w http.ResponseWriter, r *http.Request ) {
    templ.Handler(Page(r.Context())).ServeHTTP(w, r)
  })
  withMiddleware := Middleware(h)
  http.Handle("/", withMiddleware)
  http.ListenAndServe(":8080", h)
}
```

I think that the implicit context pattern is quite reasonable for the use case. It's a little bit magic, but it's a pattern that's very familiar to React developers.

Personally, I don't really want to have to pass `ctx` down the stack, because it is annoying. I had to update some code (updated database library) to pass context through the stack yesterday, and it took a fair amount of time.

I think the main concern about the `ctx` variable is that it would be confusing if you try to redefine it:

```go
templ Component() {
  // This would complain because ctx has already been defined.
  if ctx := "test"; ctx == "test" {
  }
}
```

So, the main question is... 1 vs 2.

* As @gungun974 says, number 1 is the simplest approach. I like simple. But... it leaves the little snag of ctx being magic.
* 2 would mean there would be no more magic contexts, but it requires more implementation work to support, and you've got to drill context through your UI layer manually.

## Test implementation of 2

Implementation of 2 requires adding a fake templ.Context() function and parsing and modifying all Go code written in templ to replace `templ.Context` in non-string contexts with the `templ_7745c5c3_Ctx` variable. However, this impacts the LSP, which would then also need to know about this transformation to the Go code, in order to update LSP positions within a Go expression.

I implemented 2 as a test over at https://github.com/a-h/templ/tree/templ_context_function without the required LSP modifications.

I also added a `test-context` unit test.

Given the input:

```go
package testcontext

type contextKey string

var contextKeyName contextKey = "name"

templ render() {
	<ul>
		<li>{ templ.Context().Value(contextKeyName).(string) }</li>
		if templ.Context().Value(contextKeyName).(string) == "test" {
			<li>the if passed</li>
		}
		if templ.Context().Value(contextKeyName).(string) != "test" {
			<li>the else if failed</li>
		} else if templ.Context().Value(contextKeyName).(string) == "test" {
			<li>the else if passed</li>
		}
	</ul>
}
```

And the calling code which sets the context...

```go
package testcontext

import (
	"context"
	_ "embed"
	"testing"

	"github.com/a-h/templ/generator/htmldiff"
)

//go:embed expected.html
var expected string

func Test(t *testing.T) {
	component := render()

	ctx := context.WithValue(context.Background(), contextKeyName, "test")

	diff, err := htmldiff.DiffCtx(ctx, component, expected)
	if err != nil {
		t.Fatal(err)
	}
	if diff != "" {
		t.Error(diff)
	}
}
```

You get the expected output:

```html
<ul>
				<li>test</li>
				<li>the if passed</li>
				<li>the else if passed</li>
</ul>
```

### Results of the test implementation

Noting this line in particuar, the explicit context version is quite long:

```go
		if templ.Context().Value(contextKeyName).(string) == "test" {
			<li>the if passed</li>
		}
```

Whereas the implicit `ctx` variable shortens it down.

```go
		if ctx.Value(contextKeyName).(string) == "test" {
			<li>the if passed</li>
		}
```

This is more pronounced with setting context (which would also require a transform to be applied to replace `templ.SetContext` with an implementation.

The explicit version is:

```go
{{ templ.SetContext(context.WithValue(templ.Context(), "key", "value")) }}
```

Whereas the implicit version would be...

```go
{{ ctx = context.WithValue(ctx, "key", "value") }}
```

# Conclusion

On the balance of things, I think that the best approach is to:

* Roll back the change to the `ctx` variable. Just leave it as `ctx`.
* Add the `test-context` unit test from the branch in the PR so that the ability to use implicit context values is not lost in subsequent releases.
* Document the availability of the implicit `ctx` variable within templ, and document how to use it in conjunction with middleware to pass information implicitly through a stack of components, along with suitable caveats on why "prop drilling" style approaches may be less prone to error.

**Comment by gungun974** [comment]
It's nice you take the time to try approach 2 but also to make this great review. It's exactly what I was thinking and saying.

I just didn't completely understand why you mention a `templ.SetContext()` since I never succeeded to create a new variable in Templ ? Did I miss something :sweat_smile: 

I'm agree on the fact it would be simpler to just revert `ctx` prefix name.   It's maybe a magical variable that pop from nowhere but when you look in generated code. It's make sense.

To prevent this kind of breaking change to comeback again, it's a great idea to make tests and more important to write this variable is accessible in templ component.

The only thing I think it could be important to mention in the doc is a remember that context is not necessary the best way to pass data but for some case like I expose here. It's great way to prevent terrible annoying props drilling.

Since those change are most redactional. I prefer to don't open a pool request since I'm too bad at written English. ||(I'm french for a reason mdr)||

Anyways thanks for this discussion. I'm happy to discuss this issue here ^^

**Comment by a-h** [comment]
Re: `templ.SetContext()`, no, you haven't missed anything. But... we'll have some syntax for arbitrary Go code within templates at some point in the near future as per https://github.com/a-h/templ/issues/109

Once that's in place, it makes sense to be able to set the context so you can add more variables to it.

```
templ Parent(name string) {
  {{ ctx = context.WithValue(ctx, "key", name) }}
  <div>
    @child(ctx){}
  </div>
}

templ Child(ctx context.Context) {
  <div>
    <button value={ ctx.Value("key") } />
  </div>
}
```

**Comment by a-h** [comment]
OK, I think that's everything sorted. Please do review the docs, and check it makes sense! Feel free to reopen etc.

**Comment by gungun974** [comment]
> OK, I think that's everything sorted. Please do review the docs, and check it makes sense! Feel free to reopen etc.

The doc is perfect !
- It's explain the problem.
- The props drilling solution.
- The `ctx` solution.
- And why this solution can cause issue.
- And also some example.

Thanks a lot's for solving this issue and make this official ^^

**Comment by jonerer** [comment]
Since I posted on the other thread I thought I'd chime in. I think reverting the ctx renaming is great, plus adding the docs is very nice. It's just what I was looking for earlier when confused about this topic (although it doesn't show up on the algolia search, maybe it's delayed?).

In my personal project I ended up implementing a typed struct to pass down. A bit more text to write, but the type safety is very nice. It makes sure no Page forgets to supply the context. Something like

```
type ViewContext struct {
	CurrentUser        *db.User
	CurrentActiveTeam  *db.Team
	SidebarContext     *SidebarContext
	PageTitle          string
}

func (vc *ViewContext) CurUserId() int32 {
	return vc.CurrentUser.ID
}
// etc...
```

But I agree that allowing both variants available makes a lot of sense. 

I would note something about the "untyped" version, for using implicit `ctx`. When outputting something related to a value that's been set in the context, I prefer to do something like `<div>{ CurUserName(ctx) }</div>`. And have "CurUserName" fish out whatever's been set in the context. Or return an empty string or so. That way you can choose whether to panic or not, and to keep code related to fishing things out from context out of the templates. And instead colocated to where the context is written to. But maybe that's just up to personal flavour.

**Comment by a-h** [comment]
Yes, there's a delay on the docs search. Algolia scrapes the site on a schedule to update it.

You're right on the use of untyped context. I implemented some authentication middleware this week, and I did exactly the same thing as you - implemented an `auth.UserGroups(ctx)` function to keep my other code tidy (and so that I could easily mock the implementation for tests).

I've documented this pattern in https://github.com/a-h/templ/commit/190e715586e2afac4fabcfb4ff50f7d8b9286130
